### ISSUE TITLE
Custom rank nested documents using fields as weights

### ISSUE THREAD

**Issue by petr-tik** [quickwit-oss/tantivy#611]
Below is the description of an application I would like to build. 

I think tantivy is a great match, so I have outlined a combination of feature requests (nested documents) and ask-for-help about current features (custom scoring, using field values as weights)

# Application
Indexes students and their properties including reviews of courses (which have start and end dates) and allows to search for keywords in courses ranking by a non-linear combination of recency and duration (end_date - start_date).  

**Doc structure**
Student is the top-level document. A list of students is below in JSON format. 

```json
{
  "students": [
    {
      "name": "Bob",
      "term_feedback": [
        {
          "term_start": "$DATE",
          "term_end": "$DATE",
          "comments": "Loved single variable calculus, found the descriptions hard to follow, enjoyed the exercises"
        },
        {
          "term_start": "$DATE",
          "term_end": "$DATE",
          "comments": "Struggled with multi-variable calculus, found the exam really hard"
        }
      ]
    },
    {
      "name": "Alice",
      "term_feedback": [
        {
          "term_start": "$DATE",
          "term_end": "$DATE",
          "comments": "Did lots of calculus, boring lectures"
        },
        {
          "term_start": "$DATE",
          "term_end": "$DATE",
          "comments": "Too easy, have done this before"
        }
      ]
    }
  ]
}
```

**Numbers**
10,000 documents/student, each student might have up to 20 course comments, every course comment is < 1000 words.

**Describe the solution you'd like**
Ability to index nested documents, running Phrase or Term queries and retrieve parent documents with custom scoring.

**Queries**
Look for keywords in 
term_feedback.comments[] and return students’ names

**Rank**

- Use the dates as weights. 
- More recent ranks higher than further in the past. 
- Dynamically calculate duration as number of months from dates. If the keywords are found in consecutive term_feedbacks, join term_feedback and their durations (term_start of earliest – term_end of latest). 
- Use duration and recency in a non-linear formula (TBC).

## What's missing

- [ ] Nested documents - indexing and querying, keeping the parent/child relationships
- [ ] Query interface that can run Term and Phrase queries on nested documents, retrieve parent documents
- [ ] Custom scorer - AFAIK, tantivy already has enough APIs to enable that. I would need to write own code hard-coding weights for fields + dynamically calculating duration on retrieval time to use for ranking. 

**Alternatives**
I have been reviewing the prospects of building a similar pipeline using Lucene/Solr. Lucene has support for nested documents (indexing and querying) and custom ranking functions.

As a contributor to tantivy, I would rather bring tantivy to feature parity and use it. However, time to deliver is a consideration and might trump this.

**Comment by fulmicoton** [comment]
Why close?

**Comment by petr-tik** [comment]
i went away on that day and missed your notification. 

In hindsight, this feels like a spec for a search product and not an actionable ticket for the underlying library
